feature/security/gnsi/credentialz/tests/host_certificates/host_certificates_test.go

```diff
@@ -0,0 +1,308 @@
+// limitations under the License.
+
+package host_certificates_test
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/openconfig/ondatra/gnmi"
+
+	"github.com/openconfig/featureprofiles/internal/fptest"
+	"github.com/openconfig/gnsi/credentialz"
+	tpb "github.com/openconfig/kne/proto/topo"
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/binding"
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	dutPublicKeyFilename     = "dut.pub"
+	dutCertFilename          = "dut-cert.pub"
+	hostCertificateVersion   = "v1.0"
+	hostCertificateCreatedOn = 1705962293
+)
+
+func TestMain(m *testing.M) {
+	fptest.RunTests(m)
+}
+
+func prepareCaAndHostKeys(t *testing.T, dir string) {
+	caCmd := exec.Command(
+		"ssh-keygen",
+		"-t", "ed25519",
+		"-f", "dut",
+		"-C", "dut",
+		"-q", "-N", "", // quiet, empty passphrase
+	)
+	caCmd.Dir = dir
+
+	err := caCmd.Run()
+	if err != nil {
+		t.Fatalf("failed generating ca key pair, error: %s", err)
+	}
+
+	dutCmd := exec.Command(
+		"ssh-keygen",
+		"-t", "ed25519",
+		"-f", "ca",
+		"-C", "ca",
+		"-q", "-N", "",
+	)
+	dutCmd.Dir = dir
+
+	err = dutCmd.Run()
+	if err != nil {
+		t.Fatalf("failed generating dut key pair, error: %s", err)
+	}
+}
+
+func rotateHostParameters(
+	t *testing.T,
+	dut *ondatra.DUTDevice,
+	artifacts []*credentialz.ServerKeysRequest_AuthenticationArtifacts,
+) {
+	credzClient := dut.RawAPIs().GNSI(t).Credentialz()
+
+	credzRotateClient, err := credzClient.RotateHostParameters(context.Background())
+	if err != nil {
+		t.Fatalf("failed fetching credentialz rotate host parameters client, error: %s", err)
+	}
+
+	request := &credentialz.RotateHostParametersRequest{
+		Request: &credentialz.RotateHostParametersRequest_ServerKeys{
+			ServerKeys: &credentialz.ServerKeysRequest{
+				AuthArtifacts: artifacts,
+				Version:       hostCertificateVersion,
+				CreatedOn:     hostCertificateCreatedOn,
+			},
+		},
+	}
+
+	err = credzRotateClient.Send(request)
+	if err != nil {
+		t.Fatalf("failed sending credentialz rotate host parameters request, error: %s", err)
+	}
+
+	_, err = credzRotateClient.Recv()
+	if err != nil {
+		t.Fatalf("failed receiving credentialz rotate host parameters request, error: %s", err)
+	}
+
+	err = credzRotateClient.Send(&credentialz.RotateHostParametersRequest{
+		Request: &credentialz.RotateHostParametersRequest_Finalize{
+			Finalize: request.GetFinalize(),
+		},
+	})
+	if err != nil {
+		t.Fatalf("failed sending credentialz rotate host parameters finalize request, error: %s", err)
+	}
+}
+
+func prepareDUTKeys(t *testing.T, dut *ondatra.DUTDevice, dir string) {
+	dutPrivateKeyContents, err := os.ReadFile(fmt.Sprintf("%s/dut", dir))
+	if err != nil {
+		t.Fatalf("failed reading host private key, error: %s", err)
+	}
+
+	rotateHostParameters(
+		t,
+		dut,
+		[]*credentialz.ServerKeysRequest_AuthenticationArtifacts{
+			{
+				PrivateKey: dutPrivateKeyContents,
```

Comment:
Why are we sending along private key only? If you are preparing DUT with an externally created keypair you'd need, well, the public key too.

Reply:
It is expected that the DUT derive public key from the private key itself. Please refer to this gnsi [PR](https://github.com/openconfig/gnsi/pull/63).